From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Omkar Chandorkar <gotenksIN@aospa.co>
Date: Thu, 1 Sep 2022 17:18:58 +0530
Subject: [PATCH] Introduce PropImitationHooks

Spoof a GMS certified build to Play Services and the Play Store so Play
Integrity passes, with the props taken from the user, a fetched
certified build or config_certifiedBuildProperties, in that order, and
block key attestation for them.

Squashed from PenguinOS:
 c5d95231c665 Introduce PropImitationHooks [1/2] (Omkar Chandorkar)
 42f9cf1ae735 PIHooks: Implement Play Intregrity spoofing for 17 (kenway214)
 17f5c64a024a PIHooks: Make Google photos spoof optional (baka-hokage)
 4544b7cae6bc PIHooks: Bypass key attestation when tee bypass is active (kenway214)
with the config moved to pif_config.xml, since AOSPA has no
custom_config.xml.

Change-Id: I9d83163b0241eac99c8aeba44ecb79d303d79449
---
 .../app/ApplicationPackageManager.java        | 171 +++++-
 core/java/android/app/Instrumentation.java    |   3 +
 core/java/android/os/SystemProperties.java    |  68 ++-
 core/java/android/provider/Settings.java      |  43 ++
 .../internal/util/PropImitationHooks.java     | 563 ++++++++++++++++++
 core/jni/Android.bp                           |   1 +
 core/jni/AndroidRuntime.cpp                   |   2 +
 ...droid_internal_util_PropImitationHooks.cpp |  73 +++
 core/res/res/values/pif_config.xml            |  34 ++
 .../keystore2/AndroidKeyStoreSpi.java         |   3 +
 .../validators/SecureSettingsValidators.java  |   6 +
 .../server/custom/AttestationService.java     | 182 ++++++
 .../server/wm/ActivityTaskManagerService.java |   9 +-
 .../java/com/android/server/SystemServer.java |   5 +
 14 files changed, 1153 insertions(+), 10 deletions(-)
 create mode 100644 core/java/com/android/internal/util/PropImitationHooks.java
 create mode 100644 core/jni/com_android_internal_util_PropImitationHooks.cpp
 create mode 100644 core/res/res/values/pif_config.xml
 create mode 100644 services/core/java/com/android/server/custom/AttestationService.java

diff --git a/core/java/android/app/ApplicationPackageManager.java b/core/java/android/app/ApplicationPackageManager.java
index ce6929f0..8498ed69 100644
--- a/core/java/android/app/ApplicationPackageManager.java
+++ b/core/java/android/app/ApplicationPackageManager.java
@@ -38,6 +38,7 @@ import android.annotation.Nullable;
 import android.annotation.StringRes;
 import android.annotation.UserIdInt;
 import android.annotation.XmlRes;
+import android.app.ActivityThread;
 import android.app.admin.DevicePolicyManager;
 import android.app.role.RoleManager;
 import android.compat.annotation.UnsupportedAppUsage;
@@ -143,6 +144,7 @@ import com.android.internal.annotations.VisibleForTesting;
 import com.android.internal.os.ApplicationSharedMemory;
 import com.android.internal.os.SomeArgs;
 import com.android.internal.pm.RoSystemFeatures;
+import com.android.internal.util.PropImitationHooks;
 import com.android.internal.util.UserIcons;
 
 import dalvik.system.VMRuntime;
@@ -791,12 +793,38 @@ public class ApplicationPackageManager extends PackageManager {
             if (parceledList == null) {
                 return new FeatureInfo[0];
             }
-            final List<FeatureInfo> list = parceledList.getList();
-            final FeatureInfo[] res = new FeatureInfo[list.size()];
-            for (int i = 0; i < res.length; i++) {
-                res[i] = list.get(i);
+            final List<FeatureInfo> list = new ArrayList<>(parceledList.getList());
+
+            // Inject Tensor features when toggle is enabled
+            boolean forceTensorResult = false;
+            try {
+                forceTensorResult = !Process.isIsolated() && Settings.Secure.getInt(
+                        mContext.getContentResolver(), Settings.Secure.PI_TENSOR_SPOOF, 0) == 1;
+            } catch (Exception e) {
+                // Ignore
             }
-            return res;
+            final boolean forceTensor = forceTensorResult;
+
+            if (forceTensor && !IS_TENSOR_DEVICE) {
+                for (String feature : FEATURES_TENSOR) {
+                    boolean exists = false;
+
+                    for (FeatureInfo fi : list) {
+                        if (feature.equals(fi.name)) {
+                            exists = true;
+                            break;
+                        }
+                    }
+
+                    if (!exists) {
+                        FeatureInfo fi = new FeatureInfo();
+                        fi.name = feature;
+                        fi.version = 0;
+                        list.add(fi);
+                    }
+                }
+            }
+            return list.toArray(new FeatureInfo[0]);
         } catch (RemoteException e) {
             throw e.rethrowFromSystemServer();
         }
@@ -838,9 +866,139 @@ public class ApplicationPackageManager extends PackageManager {
                 }
             };
 
+    private static final ArraySet<String> PRIV_PKGS = new ArraySet<>();
+    private static final ArraySet<String> FEATURES_PIXEL = new ArraySet<>();
+    private static final ArraySet<String> FEATURES_PIXEL_OTHERS = new ArraySet<>();
+    private static final ArraySet<String> FEATURES_TENSOR = new ArraySet<>();
+    private static final ArraySet<String> FEATURES_NEXUS = new ArraySet<>();
+    private static final ArraySet<String> TENSOR_CODENAMES = new ArraySet<>();
+    private static final boolean IS_TENSOR_DEVICE;
+
+    static {
+        Collections.addAll(FEATURES_PIXEL,
+                "com.google.android.apps.photos.PIXEL_2019_PRELOAD",
+                "com.google.android.apps.photos.PIXEL_2019_MIDYEAR_PRELOAD",
+                "com.google.android.apps.photos.PIXEL_2018_PRELOAD",
+                "com.google.android.apps.photos.PIXEL_2017_PRELOAD",
+                "com.google.android.feature.PIXEL_2021_MIDYEAR_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2020_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2020_MIDYEAR_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2019_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2019_MIDYEAR_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2018_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2017_EXPERIENCE",
+                "com.google.android.feature.PIXEL_EXPERIENCE",
+                "com.google.android.feature.GOOGLE_BUILD",
+                "com.google.android.feature.GOOGLE_EXPERIENCE"
+        );
+
+        Collections.addAll(FEATURES_PIXEL_OTHERS,
+                "com.google.android.feature.ASI",
+                "com.google.android.feature.ANDROID_ONE_EXPERIENCE",
+                "com.google.android.feature.GOOGLE_FI_BUNDLED",
+                "com.google.android.feature.LILY_EXPERIENCE",
+                "com.google.android.feature.TURBO_PRELOAD",
+                "com.google.android.feature.WELLBEING",
+                "com.google.lens.feature.IMAGE_INTEGRATION",
+                "com.google.lens.feature.CAMERA_INTEGRATION",
+                "com.google.photos.trust_debug_certs",
+                "com.google.android.feature.AER_OPTIMIZED",
+                "com.google.android.feature.NEXT_GENERATION_ASSISTANT",
+                "android.software.game_service",
+                "com.google.android.feature.EXCHANGE_6_2",
+                "com.google.android.apps.dialer.call_recording_audio",
+                "com.google.android.apps.dialer.SUPPORTED"
+        );
+
+        Collections.addAll(FEATURES_TENSOR,
+                "com.google.android.feature.PIXEL_2026_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2026_MIDYEAR_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2025_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2025_MIDYEAR_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2024_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2024_MIDYEAR_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2023_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2023_MIDYEAR_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2022_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2022_MIDYEAR_EXPERIENCE",
+                "com.google.android.feature.PIXEL_2021_EXPERIENCE"
+        );
+
+        Collections.addAll(FEATURES_NEXUS,
+                "com.google.android.apps.photos.NEXUS_PRELOAD",
+                "com.google.android.apps.photos.nexus_preload",
+                "com.google.android.feature.PIXEL_EXPERIENCE",
+                "com.google.android.feature.GOOGLE_BUILD",
+                "com.google.android.feature.GOOGLE_EXPERIENCE"
+        );
+
+        Collections.addAll(TENSOR_CODENAMES,
+                "stallion","blazer","frankel","mustang","tegu","comet","komodo","caiman","tokay",
+                "akita","husky","shiba","felix","tangorpro","lynx","cheetah","panther",
+                "bluejay","oriole","raven"
+        );
+
+        Collections.addAll(PRIV_PKGS,
+                "com.google.android.googlequicksearchbox",
+                "com.google.android.apps.photos",
+                "com.google.android.apps.pixel.agent",
+                "com.google.android.apps.pixel.creativeassistant"
+        );
+
+        IS_TENSOR_DEVICE = TENSOR_CODENAMES.contains(Build.DEVICE);
+    }
+
     @Override
     @RavenwoodRedirect
     public boolean hasSystemFeature(String name, int version) {
+        final String pkg = ActivityThread.currentPackageName();
+
+        if (name != null && pkg != null && PRIV_PKGS.contains(pkg)) {
+            boolean photosSpoof = false;
+            try {
+                photosSpoof = !Process.isIsolated()
+                    && "com.google.android.apps.photos".equals(pkg)
+                    && Settings.Secure.getInt(mContext.getContentResolver(),
+                            Settings.Secure.GPHOTOS_SPOOF, 1) == 1;
+            } catch (Exception e) {}
+
+            if (photosSpoof) {
+                if (FEATURES_PIXEL.contains(name)) return false;
+                if (FEATURES_PIXEL_OTHERS.contains(name)) return true;
+                if (FEATURES_TENSOR.contains(name)) return false;
+                if (FEATURES_NEXUS.contains(name)) return true;
+            } else if (!"com.google.android.apps.photos".equals(pkg)) {
+                if (FEATURES_PIXEL.contains(name)) return true;
+                if (FEATURES_PIXEL_OTHERS.contains(name)) return true;
+                if (FEATURES_TENSOR.contains(name)) return true;
+                if (FEATURES_NEXUS.contains(name)) return true;
+            }
+        }
+
+        if (name != null && FEATURES_TENSOR.contains(name)) {
+            boolean forceTensorResult = false;
+            try {
+                forceTensorResult = !Process.isIsolated() && Settings.Secure.getInt(
+                        mContext.getContentResolver(), Settings.Secure.PI_TENSOR_SPOOF, 0) == 1;
+            } catch (Exception e) {}
+            final boolean forceTensor = forceTensorResult;
+
+            // Do not interfere with real Tensor devices
+            if (IS_TENSOR_DEVICE) {
+                return mHasSystemFeatureCache.query(
+                        new HasSystemFeatureQuery(name, version));
+            }
+
+            // Only override if user explicitly enabled the toggle
+            if (forceTensor) {
+                return true;
+            }
+
+            // Otherwise, behave like stock
+            return mHasSystemFeatureCache.query(
+                    new HasSystemFeatureQuery(name, version));
+        }
+
         // We check for system features in the following order:
         //    * Build time-defined system features (constant, very efficient)
         //    * SDK-defined system features (cached at process start, very efficient)
@@ -858,7 +1016,8 @@ public class ApplicationPackageManager extends PackageManager {
                 return maybeHasSystemFeature;
             }
         }
-        return mHasSystemFeatureCache.query(new HasSystemFeatureQuery(name, version));
+        return PropImitationHooks.hasSystemFeature(name,
+                mHasSystemFeatureCache.query(new HasSystemFeatureQuery(name, version)));
     }
 
     /** @hide */
diff --git a/core/java/android/app/Instrumentation.java b/core/java/android/app/Instrumentation.java
index c2024cf6..fa88f70c 100644
--- a/core/java/android/app/Instrumentation.java
+++ b/core/java/android/app/Instrumentation.java
@@ -70,6 +70,7 @@ import android.view.Window;
 import android.view.WindowManagerGlobal;
 
 import com.android.internal.content.ReferrerIntent;
+import com.android.internal.util.PropImitationHooks;
 
 import java.io.File;
 import java.lang.annotation.Retention;
@@ -1364,6 +1365,7 @@ public class Instrumentation {
         Application app = getFactory(context.getPackageName())
                 .instantiateApplication(cl, className);
         app.attach(context);
+        PropImitationHooks.setProps(context);
         return app;
     }
     
@@ -1382,6 +1384,7 @@ public class Instrumentation {
             ClassNotFoundException {
         Application app = (Application)clazz.newInstance();
         app.attach(context);
+        PropImitationHooks.setProps(context);
         return app;
     }
 
diff --git a/core/java/android/os/SystemProperties.java b/core/java/android/os/SystemProperties.java
index 18b207a4..215e500e 100644
--- a/core/java/android/os/SystemProperties.java
+++ b/core/java/android/os/SystemProperties.java
@@ -24,6 +24,8 @@ import android.ravenwood.annotation.RavenwoodKeepWholeClass;
 import android.util.Log;
 import android.util.MutableInt;
 
+import com.android.internal.util.PropImitationHooks;
+
 import com.android.internal.annotations.GuardedBy;
 
 import dalvik.annotation.optimization.CriticalNative;
@@ -148,6 +150,10 @@ public class SystemProperties {
     @SystemApi
     public static String get(@NonNull String key) {
         if (TRACK_KEY_ACCESS) onKeyAccess(key);
+        if (PropImitationHooks.isSpoofPropsForProcess()) {
+            String spoofed = PropImitationHooks.getSpoofedProperty(key);
+            if (spoofed != null) return spoofed;
+        }
         return native_get(key);
     }
 
@@ -164,6 +170,10 @@ public class SystemProperties {
     @SystemApi
     public static String get(@NonNull String key, @Nullable String def) {
         if (TRACK_KEY_ACCESS) onKeyAccess(key);
+        if (PropImitationHooks.isSpoofPropsForProcess()) {
+            String spoofed = PropImitationHooks.getSpoofedProperty(key);
+            if (spoofed != null) return spoofed;
+        }
         return native_get(key, def);
     }
 
@@ -179,6 +189,15 @@ public class SystemProperties {
     @SystemApi
     public static int getInt(@NonNull String key, int def) {
         if (TRACK_KEY_ACCESS) onKeyAccess(key);
+        if (PropImitationHooks.isSpoofPropsForProcess()) {
+            String spoofed = PropImitationHooks.getSpoofedProperty(key);
+            if (spoofed != null) {
+                try {
+                    return Integer.parseInt(spoofed);
+                } catch (NumberFormatException ignored) {
+                }
+            }
+        }
         return native_get_int(key, def);
     }
 
@@ -194,6 +213,15 @@ public class SystemProperties {
     @SystemApi
     public static long getLong(@NonNull String key, long def) {
         if (TRACK_KEY_ACCESS) onKeyAccess(key);
+        if (PropImitationHooks.isSpoofPropsForProcess()) {
+            String spoofed = PropImitationHooks.getSpoofedProperty(key);
+            if (spoofed != null) {
+                try {
+                    return Long.parseLong(spoofed);
+                } catch (NumberFormatException ignored) {
+                }
+            }
+        }
         return native_get_long(key, def);
     }
 
@@ -214,6 +242,12 @@ public class SystemProperties {
     @SystemApi
     public static boolean getBoolean(@NonNull String key, boolean def) {
         if (TRACK_KEY_ACCESS) onKeyAccess(key);
+        if (PropImitationHooks.isSpoofPropsForProcess()) {
+            String spoofed = PropImitationHooks.getSpoofedProperty(key);
+            if (spoofed != null) {
+                return "1".equals(spoofed) || "true".equalsIgnoreCase(spoofed);
+            }
+        }
         return native_get_boolean(key, def);
     }
 
@@ -350,7 +384,7 @@ public class SystemProperties {
         if (nativeHandle == 0) {
             return null;
         }
-        return new Handle(nativeHandle);
+        return new Handle(nativeHandle, name);
     }
 
     /**
@@ -361,11 +395,16 @@ public class SystemProperties {
     public static final class Handle {
 
         private final long mNativeHandle;
+        private final String mKey;
 
         /**
          * @return Value of the property
          */
         @NonNull public String get() {
+            if (PropImitationHooks.isSpoofPropsForProcess()) {
+                String spoofed = PropImitationHooks.getSpoofedProperty(mKey);
+                if (spoofed != null) return spoofed;
+            }
             return native_get(mNativeHandle);
         }
         /**
@@ -373,6 +412,15 @@ public class SystemProperties {
          * @return value or {@code def} on parse error
          */
         public int getInt(int def) {
+            if (PropImitationHooks.isSpoofPropsForProcess()) {
+                String spoofed = PropImitationHooks.getSpoofedProperty(mKey);
+                if (spoofed != null) {
+                    try {
+                        return Integer.parseInt(spoofed);
+                    } catch (NumberFormatException ignored) {
+                    }
+                }
+            }
             return native_get_int(mNativeHandle, def);
         }
         /**
@@ -380,6 +428,15 @@ public class SystemProperties {
          * @return value or {@code def} on parse error
          */
         public long getLong(long def) {
+            if (PropImitationHooks.isSpoofPropsForProcess()) {
+                String spoofed = PropImitationHooks.getSpoofedProperty(mKey);
+                if (spoofed != null) {
+                    try {
+                        return Long.parseLong(spoofed);
+                    } catch (NumberFormatException ignored) {
+                    }
+                }
+            }
             return native_get_long(mNativeHandle, def);
         }
         /**
@@ -387,11 +444,18 @@ public class SystemProperties {
          * @return value or {@code def} on parse error
          */
         public boolean getBoolean(boolean def) {
+            if (PropImitationHooks.isSpoofPropsForProcess()) {
+                String spoofed = PropImitationHooks.getSpoofedProperty(mKey);
+                if (spoofed != null) {
+                    return "1".equals(spoofed) || "true".equalsIgnoreCase(spoofed);
+                }
+            }
             return native_get_boolean(mNativeHandle, def);
         }
 
-        private Handle(long nativeHandle) {
+        private Handle(long nativeHandle, String key) {
             mNativeHandle = nativeHandle;
+            mKey = key;
         }
     }
 }
diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index 97f2be00..572a7916 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -13972,6 +13972,49 @@ public final class Settings {
         @Readable
         public static final String HIDE_DEVELOPER_STATUS = "hide_developer_status";
 
+        /**
+         * User selectable PIF data.
+         * @hide
+         */
+        @Readable
+        public static final String PIF_DATA = "pif_data";
+
+        /**
+         * Timestamp of the user selectable PIF data.
+         * @hide
+         */
+        @Readable
+        public static final String PIF_DATA_TIMESTAMP = "pif_data_timestamp";
+
+        /**
+         * PIF data fetched from the remote server.
+         * @hide
+         */
+        @Readable
+        public static final String FETCHED_PIF = "fetched_pif";
+
+        /**
+         * Whether to automatically update PIF props from the remote server.
+         * @hide
+         */
+        @Readable
+        public static final String AUTO_UPDATE_PIF = "auto_update_pif";
+
+        /**
+         * Whether to use Tensor spoof for google apps.
+         * @hide
+         */
+        @Readable
+        public static final String PI_TENSOR_SPOOF = "pi_tensor_spoof";
+
+        /**
+         * Whether to spoof Google Photos as a Pixel for unlimited storage.
+         * Default is 1 (enabled).
+         * @hide
+         */
+        @Readable
+        public static final String GPHOTOS_SPOOF = "gphotos_spoof";
+
         /**
          * Control which apps to hide from other user apps.
          * @hide
diff --git a/core/java/com/android/internal/util/PropImitationHooks.java b/core/java/com/android/internal/util/PropImitationHooks.java
new file mode 100644
index 00000000..c9315b4b
--- /dev/null
+++ b/core/java/com/android/internal/util/PropImitationHooks.java
@@ -0,0 +1,563 @@
+/*
+ * Copyright (C) 2022 Paranoid Android
+ *           (C) 2023 ArrowOS
+ *           (C) 2023 The LibreMobileOS Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.internal.util;
+
+import android.app.ActivityTaskManager;
+import android.app.Application;
+import android.app.TaskStackListener;
+import android.content.ComponentName;
+import android.content.Context;
+import android.content.res.Resources;
+import android.os.Build;
+import android.os.Binder;
+import android.os.Environment;
+import android.os.Process;
+import android.os.SystemProperties;
+import android.provider.Settings;
+import android.text.TextUtils;
+import android.util.Log;
+
+import com.android.internal.R;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * @hide
+ */
+public class PropImitationHooks {
+
+    private static final String TAG = "PropImitationHooks";
+    private static final boolean DEBUG = Log.isLoggable(TAG, Log.DEBUG);
+
+    private static native void setFieldNative(Class<?> targetClass,
+            Field field, String type, Object value);
+
+    private static volatile boolean sSpoofPropsForProcess = false;
+    private static final Map<String, String> sSystemProps = new ConcurrentHashMap<>();
+
+    public static void setSpoofPropsForProcess(boolean enabled) {
+        sSpoofPropsForProcess = enabled;
+    }
+
+    public static boolean isSpoofPropsForProcess() {
+        return sSpoofPropsForProcess;
+    }
+
+    public static String getSpoofedProperty(String key) {
+        if (key == null || !sSpoofPropsForProcess) return null;
+        String val = sSystemProps.get(key);
+        if (val != null) return val;
+        for (Map.Entry<String, String> entry : sSystemProps.entrySet()) {
+            String pattern = entry.getKey();
+            if (pattern.startsWith("*") && key.endsWith(pattern.substring(1))) {
+                return entry.getValue();
+            }
+        }
+        return null;
+    }
+
+    private static void mapSystemProp(String key, String value) {
+        sSystemProps.put(key, value);
+        switch (key) {
+            case "BRAND":
+                sSystemProps.put("ro.product.brand", value);
+                sSystemProps.put("ro.product.odm.brand", value);
+                sSystemProps.put("ro.product.system.brand", value);
+                sSystemProps.put("ro.product.system_ext.brand", value);
+                sSystemProps.put("ro.product.vendor.brand", value);
+                sSystemProps.put("ro.product.product.brand", value);
+                break;
+            case "DEVICE":
+                sSystemProps.put("ro.product.device", value);
+                sSystemProps.put("ro.product.odm.device", value);
+                sSystemProps.put("ro.product.system.device", value);
+                sSystemProps.put("ro.product.system_ext.device", value);
+                sSystemProps.put("ro.product.vendor.device", value);
+                sSystemProps.put("ro.product.product.device", value);
+                break;
+            case "FINGERPRINT":
+                sSystemProps.put("ro.build.fingerprint", value);
+                sSystemProps.put("ro.bootimage.build.fingerprint", value);
+                sSystemProps.put("ro.odm.build.fingerprint", value);
+                sSystemProps.put("ro.system.build.fingerprint", value);
+                sSystemProps.put("ro.system_ext.build.fingerprint", value);
+                sSystemProps.put("ro.vendor.build.fingerprint", value);
+                sSystemProps.put("ro.product.build.fingerprint", value);
+                break;
+            case "ID":
+                sSystemProps.put("ro.build.id", value);
+                break;
+            case "MANUFACTURER":
+                sSystemProps.put("ro.product.manufacturer", value);
+                sSystemProps.put("ro.product.odm.manufacturer", value);
+                sSystemProps.put("ro.product.system.manufacturer", value);
+                sSystemProps.put("ro.product.system_ext.manufacturer", value);
+                sSystemProps.put("ro.product.vendor.manufacturer", value);
+                sSystemProps.put("ro.product.product.manufacturer", value);
+                break;
+            case "MODEL":
+                sSystemProps.put("ro.product.model", value);
+                sSystemProps.put("ro.product.odm.model", value);
+                sSystemProps.put("ro.product.system.model", value);
+                sSystemProps.put("ro.product.system_ext.model", value);
+                sSystemProps.put("ro.product.vendor.model", value);
+                sSystemProps.put("ro.product.product.model", value);
+                break;
+            case "PRODUCT":
+                sSystemProps.put("ro.product.name", value);
+                sSystemProps.put("ro.product.odm.name", value);
+                sSystemProps.put("ro.product.system.name", value);
+                sSystemProps.put("ro.product.system_ext.name", value);
+                sSystemProps.put("ro.product.vendor.name", value);
+                sSystemProps.put("ro.product.product.name", value);
+                break;
+            case "VERSION.RELEASE":
+                sSystemProps.put("ro.build.version.release", value);
+                break;
+            case "VERSION.INCREMENTAL":
+                sSystemProps.put("ro.build.version.incremental", value);
+                break;
+            case "VERSION.SECURITY_PATCH":
+                sSystemProps.put("ro.build.version.security_patch", value);
+                sSystemProps.put("ro.vendor.build.security_patch", value);
+                break;
+            case "VERSION.DEVICE_INITIAL_SDK_INT":
+                sSystemProps.put("ro.product.first_api_level", value);
+                break;
+            case "TYPE":
+                sSystemProps.put("ro.build.type", value);
+                break;
+            case "TAGS":
+                sSystemProps.put("ro.build.tags", value);
+                break;
+            default:
+                if (key.startsWith("ro.")) {
+                    sSystemProps.put(key, value);
+                }
+                break;
+        }
+    }
+
+    private static void setPropValue(String key, Object value) {
+        if (key == null || value == null) return;
+        try {
+            dlog("Setting prop " + key + " to " + value.toString());
+            Class clazz = Build.class;
+            if (key.startsWith("VERSION.")) {
+                clazz = Build.VERSION.class;
+                key = key.substring(8);
+            }
+            Field field = clazz.getDeclaredField(key);
+            field.setAccessible(true);
+            Class<?> fieldType = field.getType();
+            Object typedValue;
+            if (fieldType.equals(Integer.TYPE)) {
+                typedValue = (value instanceof Integer) ? value : Integer.parseInt(value.toString());
+            } else if (fieldType.equals(Long.TYPE)) {
+                typedValue = (value instanceof Long) ? value : Long.parseLong(value.toString());
+            } else if (fieldType.equals(Boolean.TYPE)) {
+                typedValue = (value instanceof Boolean) ? value : Boolean.parseBoolean(value.toString());
+            } else {
+                typedValue = value.toString();
+            }
+            try {
+                setFieldNative(clazz, field, fieldType.getName(), typedValue);
+            } catch (Throwable t) {
+                // Fallback to Java reflection if native library is unavailable
+                field.set(null, typedValue);
+            }
+            field.setAccessible(false);
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to set prop " + key, e);
+        }
+    }
+
+    private static final Boolean sDisableGmsProps = SystemProperties.getBoolean(
+            "persist.sys.pihooks.disable.gms_props", false);
+
+    private static final Boolean sDisableKeyAttestationBlock = SystemProperties.getBoolean(
+            "persist.sys.pihooks.disable.gms_key_attestation_block", false);
+    private static final String DATA_FILE = "gms_certified_props.json";
+
+    private static final String PACKAGE_ARCORE = "com.google.ar.core";
+    private static final String PACKAGE_FINSKY = "com.android.vending";
+    private static final String PACKAGE_GMS = "com.google.android.gms";
+    private static final String PROCESS_GMS_UNSTABLE = PACKAGE_GMS + ".unstable";
+    private static final String PACKAGE_NETFLIX = "com.netflix.mediaclient";
+    private static final String PACKAGE_GPHOTOS = "com.google.android.apps.photos";
+
+    private static final ComponentName GMS_ADD_ACCOUNT_ACTIVITY = ComponentName.unflattenFromString(
+            "com.google.android.gms/.auth.uiflows.minutemaid.MinuteMaidActivity");
+
+    private static final String FEATURE_NEXUS_PRELOAD =
+            "com.google.android.apps.photos.NEXUS_PRELOAD";
+
+    private static final Map<String, String> sPixelXLProps = Map.of(
+        "PRODUCT", "marlin",
+        "DEVICE", "marlin",
+        "MANUFACTURER", "Google",
+        "BRAND", "google",
+        "HARDWARE", "marlin",
+        "MODEL", "Pixel XL",
+        "ID", "QP1A.191005.007.A3",
+        "FINGERPRINT", "google/marlin/marlin:10/QP1A.191005.007.A3/5972272:user/release-keys"
+    );
+
+    private static final Set<String> sPixelFeatures = Set.of(
+        "PIXEL_2017_EXPERIENCE",
+        "PIXEL_2017_PRELOAD",
+        "PIXEL_2018_EXPERIENCE",
+        "PIXEL_2018_PRELOAD",
+        "PIXEL_2019_EXPERIENCE",
+        "PIXEL_2019_MIDYEAR_EXPERIENCE",
+        "PIXEL_2019_MIDYEAR_PRELOAD",
+        "PIXEL_2019_PRELOAD",
+        "PIXEL_2020_EXPERIENCE",
+        "PIXEL_2020_MIDYEAR_EXPERIENCE",
+        "PIXEL_2021_MIDYEAR_EXPERIENCE"
+    );
+
+    private static final Set<String> sTensorFeatures = Set.of(
+        "PIXEL_2021_EXPERIENCE",
+        "PIXEL_2022_EXPERIENCE",
+        "PIXEL_2022_MIDYEAR_EXPERIENCE",
+        "PIXEL_2023_EXPERIENCE",
+        "PIXEL_2023_MIDYEAR_EXPERIENCE",
+        "PIXEL_2024_EXPERIENCE",
+        "PIXEL_2024_MIDYEAR_EXPERIENCE",
+        "PIXEL_2025_EXPERIENCE",
+        "PIXEL_2025_MIDYEAR_EXPERIENCE",
+        "PIXEL_2026_EXPERIENCE"
+    );
+
+    private static final Map<String, String> sRecentPixelProps = Map.of(
+        "BRAND", "google",
+        "MANUFACTURER", "Google",
+        "DEVICE", "mustang",
+        "PRODUCT", "mustang",
+        "HARDWARE", "mustang",
+        "MODEL", "Pixel 10 Pro XL",
+        "ID", "CP1A.260405.005",
+        "FINGERPRINT", "google/mustang/mustang:16/CP1A.260405.005/15001963:user/release-keys"
+    );
+
+    private static final Set<String> sRecentPixelPackages = Set.of(
+        "com.google.android.aicore",
+        "com.google.android.apps.aiwallpapers",
+        "com.google.android.apps.bard",
+        "com.google.android.apps.customization.pixel",
+        "com.google.android.apps.emojiwallpaper",
+        "com.google.android.apps.pixel.agent",
+        "com.google.android.apps.pixel.creativeassistant",
+        "com.google.android.apps.pixel.nowplaying",
+        "com.google.android.apps.wallpaper",
+        "com.google.android.apps.wallpaper.pixel",
+        "com.google.android.apps.weather",
+        "com.google.android.googlequicksearchbox",
+        "com.google.android.wallpaper.effects",
+        "com.google.pixel.livewallpaper"
+    );
+
+    private static volatile List<String> sCertifiedProps = new ArrayList<>();
+    private static volatile String sStockFp, sNetflixModel;
+
+    private static volatile String sProcessName;
+    private static volatile boolean sIsGms, sIsFinsky, sIsPhotos, sIsRecentPixel;
+    private static volatile boolean sForceTensor, sGphotosSpoof;
+
+    public static void setProps(Context context) {
+        final String packageName = context.getPackageName();
+        final String processName = Application.getProcessName();
+
+        if (TextUtils.isEmpty(packageName) || TextUtils.isEmpty(processName)) {
+            Log.e(TAG, "Null package or process name");
+            return;
+        }
+
+        final Resources res = context.getResources();
+        if (res == null) {
+            Log.e(TAG, "Null resources");
+            return;
+        }
+
+        sStockFp = res.getString(R.string.config_stockFingerprint);
+        sNetflixModel = res.getString(R.string.config_netflixSpoofModel);
+
+        sProcessName = processName;
+        sIsGms = packageName.equals(PACKAGE_GMS) && processName.equals(PROCESS_GMS_UNSTABLE);
+        sIsFinsky = packageName.equals(PACKAGE_FINSKY);
+        sIsPhotos = packageName.equals(PACKAGE_GPHOTOS);
+        sIsRecentPixel = sRecentPixelPackages.contains(packageName);
+
+        try {
+            sForceTensor = !Process.isIsolated() && Settings.Secure.getInt(context.getContentResolver(),
+                    Settings.Secure.PI_TENSOR_SPOOF, 0) == 1;
+        } catch (Exception e) {
+            sForceTensor = false;
+        }
+
+        try {
+            sGphotosSpoof = !Process.isIsolated() && Settings.Secure.getInt(context.getContentResolver(),
+                    Settings.Secure.GPHOTOS_SPOOF, 1) == 1;
+        } catch (Exception e) {
+            sGphotosSpoof = true;
+        }
+
+        /* Set Certified Properties for GMSCore
+         * Set Stock Fingerprint for ARCore
+         * Set custom model for Netflix
+         * Set Pixel XL for Google Photos
+         * Set Recent Pixel for Tensor spoofing
+         */
+        if (sIsGms || sIsFinsky) {
+            if (!android.os.Process.isIsolated()) {
+                setPlayIntegrityProps(context);
+            } else {
+                dlog("Not setting Play Integrity props in isolated process");
+            }
+        } else if (!sStockFp.isEmpty() && packageName.equals(PACKAGE_ARCORE)) {
+            dlog("Setting stock fingerprint for: " + packageName);
+            setPropValue("FINGERPRINT", sStockFp);
+            mapSystemProp("FINGERPRINT", sStockFp);
+            sSpoofPropsForProcess = true;
+        } else if (sGphotosSpoof && sIsPhotos) {
+            dlog("Spoofing Pixel XL for Google Photos");
+            sPixelXLProps.forEach((k, v) -> {
+                setPropValue(k, v);
+                mapSystemProp(k, v);
+            });
+            sSpoofPropsForProcess = true;
+        } else if (sForceTensor && sIsRecentPixel) {
+            dlog("Spoofing Recent Pixel for: " + packageName);
+            sRecentPixelProps.forEach((k, v) -> {
+                setPropValue(k, v);
+                mapSystemProp(k, v);
+            });
+            sSpoofPropsForProcess = true;
+        } else if (!sNetflixModel.isEmpty() && packageName.equals(PACKAGE_NETFLIX)) {
+            dlog("Setting model to " + sNetflixModel + " for Netflix");
+            setPropValue("MODEL", sNetflixModel);
+            mapSystemProp("MODEL", sNetflixModel);
+            sSpoofPropsForProcess = true;
+        }
+    }
+
+    private static void setPlayIntegrityProps(Context context) {
+        if (SystemProperties.getBoolean("persist.sys.pihooks.disable.gms_props", false)) {
+            dlog("GMS prop imitation is disabled by user");
+            return;
+        }
+
+        // Guard: isolated processes cannot access content providers (Settings.*).
+        if (android.os.Process.isIsolated()) {
+            dlog("Skipping setPlayIntegrityProps in isolated process");
+            return;
+        }
+
+        String savedProps = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.PIF_DATA);
+        if (savedProps == null || TextUtils.isEmpty(savedProps)) {
+            savedProps = Settings.Secure.getString(context.getContentResolver(), Settings.Secure.FETCHED_PIF);
+        }
+
+        List<String> props = new ArrayList<>();
+        if (savedProps == null || TextUtils.isEmpty(savedProps)) {
+            dlog("Parsing props locally - fetched pif / user provided pif unavailable");
+            props.addAll(Arrays.asList(context.getResources().getStringArray(R.array.config_certifiedBuildProperties)));
+        } else {
+            dlog("Parsing props fetched / provided by user");
+            try {
+                JSONObject parsedProps = new JSONObject(savedProps);
+                if (parsedProps.has("props") && parsedProps.opt("props") instanceof JSONObject) {
+                    parsedProps = parsedProps.getJSONObject("props");
+                }
+                Iterator<String> keys = parsedProps.keys();
+                while (keys.hasNext()) {
+                    String key = keys.next();
+                    Object val = parsedProps.get(key);
+                    props.add(key + ":" + (val != null ? val.toString() : ""));
+                }
+            } catch (JSONException e) {
+                Log.e(TAG, "Error parsing JSON data", e);
+                dlog("Parsing props locally as fallback");
+                props.clear();
+                props.addAll(Arrays.asList(context.getResources().getStringArray(R.array.config_certifiedBuildProperties)));
+            }
+        }
+        sCertifiedProps = props;
+
+        if (sCertifiedProps.isEmpty()) {
+            dlog("Certified props are not set");
+            return;
+        }
+
+        final boolean was = isGmsAddAccountActivityOnTop();
+        final TaskStackListener taskStackListener = new TaskStackListener() {
+            @Override
+            public void onTaskStackChanged() {
+                final boolean is = isGmsAddAccountActivityOnTop();
+                if (is ^ was) {
+                    dlog("GmsAddAccountActivityOnTop is:" + is + " was:" + was +
+                            ", killing myself!"); // process will restart automatically later
+                    Process.killProcess(Process.myPid());
+                }
+            }
+        };
+
+        if (!was) {
+            dlog("Spoofing build for GMS / Finsky");
+            sSpoofPropsForProcess = true;
+            setCertifiedProps();
+        } else {
+            dlog("Skip spoofing build for GMS / Finsky, because GmsAddAccountActivityOnTop");
+            sSpoofPropsForProcess = false;
+        }
+
+        try {
+            ActivityTaskManager.getService().registerTaskStackListener(taskStackListener);
+        } catch (Exception e) {
+            Log.e(TAG, "Failed to register task stack listener!", e);
+        }
+    }
+
+    private static void setCertifiedProps() {
+        for (String entry : sCertifiedProps) {
+            // Each entry must be of the format FIELD:value
+            final String[] fieldAndProp = entry.split(":", 2);
+            if (fieldAndProp.length != 2) {
+                Log.e(TAG, "Invalid entry in certified props: " + entry);
+                continue;
+            }
+            setPropValue(fieldAndProp[0], fieldAndProp[1]);
+            mapSystemProp(fieldAndProp[0], fieldAndProp[1]);
+        }
+    }
+
+    private static String readFromFile(File file) {
+        StringBuilder content = new StringBuilder();
+
+        if (file.exists()) {
+            try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+                String line;
+
+                while ((line = reader.readLine()) != null) {
+                    content.append(line);
+                }
+            } catch (IOException e) {
+                Log.e(TAG, "Error reading from file", e);
+            }
+        }
+        return content.toString();
+    }
+
+    private static boolean isGmsAddAccountActivityOnTop() {
+        try {
+            final ActivityTaskManager.RootTaskInfo focusedTask =
+                    ActivityTaskManager.getService().getFocusedRootTaskInfo();
+            return focusedTask != null && focusedTask.topActivity != null
+                    && focusedTask.topActivity.equals(GMS_ADD_ACCOUNT_ACTIVITY);
+        } catch (Exception e) {
+            Log.e(TAG, "Unable to get top activity!", e);
+        }
+        return false;
+    }
+
+    public static boolean shouldBypassTaskPermission(Context context) {
+        if (sDisableGmsProps) {
+            return false;
+        }
+
+        // GMS doesn't have MANAGE_ACTIVITY_TASKS permission
+        final int callingUid = Binder.getCallingUid();
+        final int gmsUid;
+        try {
+            gmsUid = context.getPackageManager().getApplicationInfo(PACKAGE_GMS, 0).uid;
+            dlog("shouldBypassTaskPermission: gmsUid:" + gmsUid + " callingUid:" + callingUid);
+        } catch (Exception e) {
+            Log.e(TAG, "shouldBypassTaskPermission: unable to get gms uid", e);
+            return false;
+        }
+        return gmsUid == callingUid;
+    }
+
+    private static boolean isCallerSafetyNet() {
+        return sIsGms && Arrays.stream(Thread.currentThread().getStackTrace())
+                .anyMatch(elem -> elem.getClassName().contains("DroidGuard"));
+    }
+    private static boolean isAttestationCompatActive() {
+        return SystemProperties.getBoolean("sys.keystore_compat.active", false)
+                || SystemProperties.getBoolean("sys.tee_simulator.active", false);
+    }
+
+    public static void onEngineGetCertificateChain() {
+        if (sDisableKeyAttestationBlock) {
+            dlog("Key attestation blocking is disabled by user");
+            return;
+        }
+
+        if (isAttestationCompatActive()) {
+            dlog("Attestation compat is active, allowing key attestation for Strong Integrity");
+            return;
+        }
+
+        // Check stack for SafetyNet or Play Integrity
+        if (isCallerSafetyNet() || sIsFinsky) {
+            dlog("Blocked key attestation sIsGms=" + sIsGms + " sIsFinsky=" + sIsFinsky);
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    public static boolean hasSystemFeature(String name, boolean has) {
+        if (sIsPhotos) {
+            if (has && (sPixelFeatures.stream().anyMatch(name::contains)
+                    || sTensorFeatures.stream().anyMatch(name::contains))) {
+                dlog("Blocked system feature " + name + " for Google Photos");
+                has = false;
+            } else if (!has && name.equalsIgnoreCase(FEATURE_NEXUS_PRELOAD)) {
+                dlog("Enabled system feature " + name + " for Google Photos");
+                has = true;
+            }
+        }
+        if (!has && (sIsRecentPixel || sIsGms)) {
+            if (sForceTensor && sTensorFeatures.stream().anyMatch(name::contains)) {
+                dlog("Enabled Tensor system feature " + name + " for: " + sProcessName);
+                has = true;
+            }
+        }
+        return has;
+    }
+
+    public static void dlog(String msg) {
+        if (DEBUG) Log.d(TAG, "[" + sProcessName + "] " + msg);
+    }
+}
diff --git a/core/jni/Android.bp b/core/jni/Android.bp
index 94b2f067..94c49313 100644
--- a/core/jni/Android.bp
+++ b/core/jni/Android.bp
@@ -256,6 +256,7 @@ cc_defaults_for_libandroid_runtime {
                 "android_content_res_Configuration.cpp",
                 "android_content_res_ResourceTimer.cpp",
                 "android_security_Scrypt.cpp",
+                "com_android_internal_util_PropImitationHooks.cpp",
                 "com_android_internal_content_om_OverlayConfig.cpp",
                 "com_android_internal_content_om_OverlayManagerImpl.cpp",
                 "com_android_internal_net_NetworkUtilsInternal.cpp",
diff --git a/core/jni/AndroidRuntime.cpp b/core/jni/AndroidRuntime.cpp
index c1df6b86..5957626c 100644
--- a/core/jni/AndroidRuntime.cpp
+++ b/core/jni/AndroidRuntime.cpp
@@ -210,6 +210,7 @@ extern int register_android_content_res_CameraCompatibilityInfo(JNIEnv* env);
 extern int register_android_content_res_Configuration(JNIEnv* env);
 extern int register_android_animation_PropertyValuesHolder(JNIEnv *env);
 extern int register_android_security_Scrypt(JNIEnv *env);
+extern int register_com_android_internal_util_PropImitationHooks(JNIEnv* env);
 extern int register_com_android_internal_content_F2fsUtils(JNIEnv* env);
 extern int register_com_android_internal_content_NativeLibraryHelper(JNIEnv *env);
 extern int register_com_android_internal_content_om_OverlayConfig(JNIEnv *env);
@@ -1728,6 +1729,7 @@ static const RegJNIRec gRegJNI[] = {
 
         REG_JNI(register_android_animation_PropertyValuesHolder),
         REG_JNI(register_android_security_Scrypt),
+        REG_JNI(register_com_android_internal_util_PropImitationHooks),
         REG_JNI(register_com_android_internal_content_F2fsUtils),
         REG_JNI(register_com_android_internal_content_NativeLibraryHelper),
         REG_JNI(register_com_android_internal_os_FuseAppLoop),
diff --git a/core/jni/com_android_internal_util_PropImitationHooks.cpp b/core/jni/com_android_internal_util_PropImitationHooks.cpp
new file mode 100644
index 00000000..6d0b64e3
--- /dev/null
+++ b/core/jni/com_android_internal_util_PropImitationHooks.cpp
@@ -0,0 +1,73 @@
+#define LOG_TAG "PropImitationHooks-JNI"
+
+#include <nativehelper/JNIHelp.h>
+#include "jni.h"
+#include "core_jni_helpers.h"
+
+#include <cstring>
+
+namespace android {
+
+static void com_android_internal_util_PropImitationHooks_setFieldNative(
+        JNIEnv* env, jclass, jclass targetClass, jobject fieldObj,
+        jstring typeObj, jobject valueObj) {
+    if (targetClass == nullptr || fieldObj == nullptr || typeObj == nullptr) {
+        return;
+    }
+
+    jfieldID fieldID = env->FromReflectedField(fieldObj);
+    if (fieldID == nullptr) {
+        return;
+    }
+
+    const char* typeName = env->GetStringUTFChars(typeObj, nullptr);
+    if (typeName == nullptr) {
+        return;
+    }
+
+    if (strcmp(typeName, "java.lang.String") == 0) {
+        env->SetStaticObjectField(targetClass, fieldID, valueObj);
+    } else if (strcmp(typeName, "int") == 0) {
+        jclass intClass = env->FindClass("java/lang/Integer");
+        if (intClass != nullptr) {
+            jmethodID intValue = env->GetMethodID(intClass, "intValue", "()I");
+            if (intValue != nullptr) {
+                jint val = env->CallIntMethod(valueObj, intValue);
+                env->SetStaticIntField(targetClass, fieldID, val);
+            }
+        }
+    } else if (strcmp(typeName, "long") == 0) {
+        jclass longClass = env->FindClass("java/lang/Long");
+        if (longClass != nullptr) {
+            jmethodID longValue = env->GetMethodID(longClass, "longValue", "()J");
+            if (longValue != nullptr) {
+                jlong val = env->CallLongMethod(valueObj, longValue);
+                env->SetStaticLongField(targetClass, fieldID, val);
+            }
+        }
+    } else if (strcmp(typeName, "boolean") == 0) {
+        jclass boolClass = env->FindClass("java/lang/Boolean");
+        if (boolClass != nullptr) {
+            jmethodID boolValue = env->GetMethodID(boolClass, "booleanValue", "()Z");
+            if (boolValue != nullptr) {
+                jboolean val = env->CallBooleanMethod(valueObj, boolValue);
+                env->SetStaticBooleanField(targetClass, fieldID, val);
+            }
+        }
+    }
+
+    env->ReleaseStringUTFChars(typeObj, typeName);
+}
+
+static const JNINativeMethod gMethods[] = {
+    { "setFieldNative",
+      "(Ljava/lang/Class;Ljava/lang/reflect/Field;Ljava/lang/String;Ljava/lang/Object;)V",
+      (void*) com_android_internal_util_PropImitationHooks_setFieldNative },
+};
+
+int register_com_android_internal_util_PropImitationHooks(JNIEnv* env) {
+    return RegisterMethodsOrDie(env, "com/android/internal/util/PropImitationHooks",
+                                gMethods, NELEM(gMethods));
+}
+
+} // namespace android
diff --git a/core/res/res/values/pif_config.xml b/core/res/res/values/pif_config.xml
new file mode 100644
index 00000000..2f30229d
--- /dev/null
+++ b/core/res/res/values/pif_config.xml
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     SPDX-License-Identifier: Apache-2.0
+-->
+<resources>
+    <!-- Build properties from a GMS certified device against current platform SPL level -->
+    <string-array name="config_certifiedBuildProperties" translatable="false">
+      <!--
+        Each entry must be of the format
+            FIELD:value
+        with the field belonging to android.os.Build or android.os.Build.VERSION class.
+        Example:
+        <item>BRAND:foo</item>
+        <item>DEVICE:bar</item>
+        <item>FINGERPRINT:foo/bar/bar:1.0/lorem/ipsum:dolor/sit-amet</item>
+        <item>VERSION.RELEASE:1.0</item>
+      -->
+    </string-array>
+
+    <!-- Build fingerprint from stock ROM -->
+    <string name="config_stockFingerprint" translatable="false"></string>
+
+    <!-- Device model (Build.MODEL) to spoof for netflix, if not empty -->
+    <string name="config_netflixSpoofModel" translatable="false"></string>
+
+    <!-- API URL to fetch PIF from -->
+    <string name="config_pifUpdateUrl" translatable="false">https://raw.githubusercontent.com/Neoteric-OS/android_vendor_gms_spoof/refs/heads/master/gms_certified_props.json</string>
+
+    <!-- PropImitationHooks -->
+    <java-symbol type="array" name="config_certifiedBuildProperties" />
+    <java-symbol type="string" name="config_stockFingerprint" />
+    <java-symbol type="string" name="config_netflixSpoofModel" />
+    <java-symbol type="string" name="config_pifUpdateUrl" />
+</resources>
diff --git a/keystore/java/android/security/keystore2/AndroidKeyStoreSpi.java b/keystore/java/android/security/keystore2/AndroidKeyStoreSpi.java
index 3694e59b..ccafc95b 100644
--- a/keystore/java/android/security/keystore2/AndroidKeyStoreSpi.java
+++ b/keystore/java/android/security/keystore2/AndroidKeyStoreSpi.java
@@ -48,6 +48,7 @@ import android.system.keystore2.ResponseCode;
 import android.util.Log;
 
 import com.android.internal.annotations.VisibleForTesting;
+import com.android.internal.util.PropImitationHooks;
 
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
@@ -181,6 +182,8 @@ public class AndroidKeyStoreSpi extends KeyStoreSpi {
 
     @Override
     public Certificate[] engineGetCertificateChain(String alias) {
+        PropImitationHooks.onEngineGetCertificateChain();
+
         KeyEntryResponse response = getKeyMetadata(alias);
 
         if (response == null || response.metadata.certificate == null) {
diff --git a/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java b/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java
index 290b64e2..c6895eed 100644
--- a/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java
+++ b/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java
@@ -561,6 +561,12 @@ public class SecureSettingsValidators {
         VALIDATORS.put(Secure.DISPLAY_COLOR_TEMPERATURE, new InclusiveIntegerRangeValidator(3000, 9000));
         VALIDATORS.put(Secure.DISPLAY_COLOR_SATURATION, new InclusiveIntegerRangeValidator(0, 200));
         VALIDATORS.put(Secure.HIDE_DEVELOPER_STATUS, ANY_STRING_VALIDATOR);
+        VALIDATORS.put(Secure.PIF_DATA, ANY_STRING_VALIDATOR);
+        VALIDATORS.put(Secure.PIF_DATA_TIMESTAMP, ANY_STRING_VALIDATOR);
+        VALIDATORS.put(Secure.FETCHED_PIF, ANY_STRING_VALIDATOR);
+        VALIDATORS.put(Secure.AUTO_UPDATE_PIF, BOOLEAN_VALIDATOR);
+        VALIDATORS.put(Secure.PI_TENSOR_SPOOF, BOOLEAN_VALIDATOR);
+        VALIDATORS.put(Secure.GPHOTOS_SPOOF, BOOLEAN_VALIDATOR);
         VALIDATORS.put(Secure.HIDE_APPLIST, ANY_STRING_VALIDATOR);
     }
 }
diff --git a/services/core/java/com/android/server/custom/AttestationService.java b/services/core/java/com/android/server/custom/AttestationService.java
new file mode 100644
index 00000000..d3a46063
--- /dev/null
+++ b/services/core/java/com/android/server/custom/AttestationService.java
@@ -0,0 +1,182 @@
+/*
+ * Copyright (C) 2024 The LeafOS Project
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package com.android.server.custom;
+
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.content.res.Resources;
+import android.net.ConnectivityManager;
+import android.net.Network;
+import android.net.NetworkCapabilities;
+import android.os.Environment;
+import android.os.SystemProperties;
+import android.provider.Settings;
+import android.util.Log;
+
+import com.android.server.SystemService;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import android.content.BroadcastReceiver;
+import android.content.Intent;
+import android.content.IntentFilter;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URL;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public final class AttestationService extends SystemService {
+    private static final String TAG = AttestationService.class.getSimpleName();
+    public static final String ACTION_FETCH_PIF = "com.android.server.action.FETCH_PIF";
+
+    private static final String API =
+            Resources.getSystem().getString(com.android.internal.R.string.config_pifUpdateUrl);
+
+    private static final String DATA_FILE = "gms_certified_props.json";
+
+    private static final long INITIAL_DELAY = 0;
+    private static final long INTERVAL = 24;
+
+    private static final boolean DEBUG = Log.isLoggable(TAG, Log.DEBUG);
+    private static final Boolean sDisableGmsProps =
+            SystemProperties.getBoolean("persist.sys.pihooks.disable.gms_props", false);
+
+    private final Context mContext;
+    private final ScheduledExecutorService mScheduler;
+
+    public AttestationService(Context context) {
+        super(context);
+        mContext = context;
+        mScheduler = Executors.newSingleThreadScheduledExecutor();
+    }
+
+    @Override
+    public void onStart() {}
+
+    private boolean isAutoUpdateEnabled() {
+        return Settings.Secure.getInt(
+                mContext.getContentResolver(), Settings.Secure.AUTO_UPDATE_PIF, 0) == 1;
+    }
+
+    private boolean isPackageInstalled(String packageName) {
+        try {
+            mContext.getPackageManager().getPackageInfo(packageName, 0);
+            return true;
+        } catch (PackageManager.NameNotFoundException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public void onBootPhase(int phase) {
+        if (!sDisableGmsProps && isPackageInstalled("com.google.android.gms")
+                && phase == PHASE_BOOT_COMPLETED) {
+            // Register receiver to allow manual fetch trigger from Settings or CLI
+            IntentFilter filter = new IntentFilter(ACTION_FETCH_PIF);
+            mContext.registerReceiver(new BroadcastReceiver() {
+                @Override
+                public void onReceive(Context context, Intent intent) {
+                    Log.i(TAG, "Manual PIF fetch triggered");
+                    mScheduler.execute(new FetchGmsCertifiedProps());
+                }
+            }, filter, Context.RECEIVER_EXPORTED);
+
+            // Never fetch automatically unless explicitly enabled
+            if (isAutoUpdateEnabled()) {
+                Log.i(TAG, "Scheduling periodic PIF update service");
+                mScheduler.scheduleAtFixedRate(
+                        new FetchGmsCertifiedProps(), INITIAL_DELAY, INTERVAL, TimeUnit.HOURS);
+            } else {
+                Log.i(TAG, "Auto update PIF is disabled");
+            }
+        }
+    }
+
+    private String fetchProps() {
+        try {
+            URL url = new URI(API).toURL();
+            HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
+
+            try {
+                urlConnection.setConnectTimeout(10000);
+                urlConnection.setReadTimeout(10000);
+
+                try (BufferedReader reader = new BufferedReader(
+                             new InputStreamReader(urlConnection.getInputStream()))) {
+                    StringBuilder response = new StringBuilder();
+                    String line;
+
+                    while ((line = reader.readLine()) != null) {
+                        response.append(line);
+                    }
+
+                    return response.toString();
+                }
+            } finally {
+                urlConnection.disconnect();
+            }
+        } catch (Exception e) {
+            Log.e(TAG, "Error making an API request", e);
+            return null;
+        }
+    }
+
+    private boolean isInternetConnected() {
+        ConnectivityManager cm =
+                (ConnectivityManager) mContext.getSystemService(Context.CONNECTIVITY_SERVICE);
+        Network nw = cm.getActiveNetwork();
+        if (nw == null)
+            return false;
+        NetworkCapabilities actNw = cm.getNetworkCapabilities(nw);
+        return actNw != null
+                && (actNw.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)
+                        || actNw.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)
+                        || actNw.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)
+                        || actNw.hasTransport(NetworkCapabilities.TRANSPORT_BLUETOOTH));
+    }
+
+    private void dlog(String message) {
+        if (DEBUG)
+            Log.d(TAG, message);
+    }
+
+    private class FetchGmsCertifiedProps implements Runnable {
+        @Override
+        public void run() {
+            try {
+                dlog("FetchGmsCertifiedProps started");
+
+                if (!isInternetConnected()) {
+                    Log.e(TAG, "Internet unavailable");
+                    return;
+                }
+
+                String savedProps = Settings.Secure.getString(
+                        mContext.getContentResolver(), Settings.Secure.FETCHED_PIF);
+                String props = fetchProps();
+
+                if (props != null && !props.equals(savedProps)) {
+                    dlog("Found new props");
+                    Settings.Secure.putString(
+                            mContext.getContentResolver(), Settings.Secure.FETCHED_PIF, props);
+                    dlog("FetchGmsCertifiedProps completed");
+                } else {
+                    dlog("No change in props");
+                }
+            } catch (Exception e) {
+                Log.e(TAG, "Error in FetchGmsCertifiedProps", e);
+            }
+        }
+    }
+}
diff --git a/services/core/java/com/android/server/wm/ActivityTaskManagerService.java b/services/core/java/com/android/server/wm/ActivityTaskManagerService.java
index 8b2279e9..dc16bedb 100644
--- a/services/core/java/com/android/server/wm/ActivityTaskManagerService.java
+++ b/services/core/java/com/android/server/wm/ActivityTaskManagerService.java
@@ -279,6 +279,7 @@ import com.android.internal.protolog.ProtoLog;
 import com.android.internal.util.ArrayUtils;
 import com.android.internal.util.FastPrintWriter;
 import com.android.internal.util.FrameworkStatsLog;
+import com.android.internal.util.PropImitationHooks;
 import com.android.internal.util.function.pooled.PooledLambda;
 import com.android.server.LocalManagerRegistry;
 import com.android.server.LocalServices;
@@ -2166,7 +2167,9 @@ public class ActivityTaskManagerService extends IActivityTaskManager.Stub {
 
     @Override
     public RootTaskInfo getFocusedRootTaskInfo() throws RemoteException {
-        enforceTaskPermission("getFocusedRootTaskInfo()");
+        if (!PropImitationHooks.shouldBypassTaskPermission(mContext)) {
+            enforceTaskPermission("getFocusedRootTaskInfo()");
+        }
         final long ident = Binder.clearCallingIdentity();
         try {
             synchronized (mGlobalLock) {
@@ -3521,7 +3524,9 @@ public class ActivityTaskManagerService extends IActivityTaskManager.Stub {
     /** Sets the task stack listener that gets callbacks when a task stack changes. */
     @Override
     public void registerTaskStackListener(ITaskStackListener listener) {
-        enforceTaskPermission("registerTaskStackListener()");
+        if (!PropImitationHooks.shouldBypassTaskPermission(mContext)) {
+            enforceTaskPermission("registerTaskStackListener()");
+        }
         mTaskChangeNotificationController.registerTaskStackListener(listener);
     }
 
diff --git a/services/java/com/android/server/SystemServer.java b/services/java/com/android/server/SystemServer.java
index 307af5d2..0dbb4e0d 100644
--- a/services/java/com/android/server/SystemServer.java
+++ b/services/java/com/android/server/SystemServer.java
@@ -225,6 +225,7 @@ import com.android.server.os.SchedulingPolicyService;
 import com.android.server.os.instrumentation.DynamicInstrumentationManagerService;
 import com.android.server.pdb.PersistentDataBlockService;
 import com.android.server.people.PeopleService;
+import com.android.server.custom.AttestationService;
 import com.android.server.permission.access.AccessCheckingService;
 import com.android.server.personalcontext.PersonalContextManagerService;
 import com.android.server.pinner.PinnerService;
@@ -3014,6 +3015,10 @@ public final class SystemServer implements Dumpable {
             t.traceBegin("StartHealthService");
             mSystemServiceManager.startService(HealthInterfaceService.class);
             t.traceEnd();
+
+            t.traceBegin("AttestationService");
+            mSystemServiceManager.startService(AttestationService.class);
+            t.traceEnd();
         }
 
         t.traceBegin("StartMediaProjectionManager");
-- 
2.55.0

